Mathlib/Algebra/Star/Positive.lean

```diff
@@ -0,0 +1,178 @@
+  Iff.rfl
+
+theorem positive_coe_set [OrderedAddCommMonoid R] : (positive R : Set R) = {x | 0 ≤ x} := rfl
+
+section NonUnitalSemiring
+
+variable [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R]
+
+theorem positive_iff {x : R} :
+    x ∈ positive R ↔ x ∈ AddSubmonoid.closure (Set.range fun s : R => star s * s) := by
+  rw [positive_def, nonneg_iff]
+
+theorem isSelfAdjoint_of_nonneg [NonUnitalSemiring R] [PartialOrder R] [StarOrderedRing R]
+    {x : R} (hx : 0 ≤ x) : IsSelfAdjoint x := by
+  rw [nonneg_iff] at hx
+  refine AddSubmonoid.closure_induction hx ?_ (isSelfAdjoint_zero R) fun _ _ => IsSelfAdjoint.add
+  rintro - ⟨s, rfl⟩
+  exact IsSelfAdjoint.star_mul_self s
+
+theorem le_iff_exists_positive {x y : R} : x ≤ y ↔ ∃ p : positive R, y = x + p := by
+  rw [le_iff']
+  simp only [Subtype.exists, exists_prop]
+  rfl
+
+@[norm_cast]
+theorem positive.val_le_val {x y : positive R} : (x : R) ≤ y ↔ x ≤ y :=
+  Iff.rfl
+
+instance positive.instCanonicallyOrderedAddMonoid : CanonicallyOrderedAddMonoid (positive R) :=
+  { inferInstanceAs (OrderedAddCommMonoid (positive R)) with
+    bot := 0
+    bot_le := Subtype.prop
+    exists_add_of_le := @fun x y (h : (x : R) ≤ y) => by
+      convert le_iff_exists_positive.mp h using 1
+      funext p
+      exact_mod_cast rfl
+    le_self_add := fun x y => show (x : R) ≤ x + y from le_iff_exists_positive.mpr ⟨y, rfl⟩ }
+
+theorem positive.isSelfAdjoint (x : positive R) : IsSelfAdjoint (x : R) :=
+  isSelfAdjoint_of_nonneg x.prop
+
+end NonUnitalSemiring
+
+namespace positive
+
+section NonUnitalRing
+
+variable [NonUnitalRing R] [PartialOrder R] [StarOrderedRing R]
+
+/-- The identity map from `positive R` to `selfAdjoint R`, as an `AddMonoidHom`. -/
+protected def selfAdjoint : positive R →+ selfAdjoint R where
+  toFun := Subtype.map id (fun _ => isSelfAdjoint_of_nonneg)
+  map_zero' := rfl
+  map_add' _ _ := rfl
+
+@[simp]
+theorem val_selfAdjoint (x : positive R) : (positive.selfAdjoint x : R) = x :=
+  rfl
+
+lemma selfAdjoint_le (x y : positive R) :
+    positive.selfAdjoint x ≤ positive.selfAdjoint y ↔ x ≤ y :=
+  Iff.rfl
+
+end NonUnitalRing
+
+section Semiring
+
+variable [Semiring R] [PartialOrder R] [StarOrderedRing R]
+
+variable (R)
+
+instance instOne : One (positive R) := ⟨⟨1, zero_le_one⟩⟩
+
+@[simp, norm_cast]
+lemma val_one : (1 : positive R) = (1 : R) :=
+  rfl
+
+variable {R}
+
+instance instNeZeroOne [Nontrivial R] : NeZero (1 : positive R) where
+  out := by rw [Ne.def, Subtype.ext_iff]; exact one_ne_zero
+
+instance instNontrivial [Nontrivial R] : Nontrivial (positive R) where
+  exists_pair_ne := ⟨1, 0, NeZero.ne 1⟩
+
+instance instNatCast : NatCast (positive R) where
+  natCast n := ⟨n, n.cast_nonneg'⟩
+
+@[simp, norm_cast]
+theorem val_natCast (n : ℕ) : ((n : positive R) : R) = n := rfl
+
+instance instAddCommMonoidWithOne : AddCommMonoidWithOne (positive R) :=
+  Subtype.val_injective.addCommMonoidWithOne _ rfl rfl (fun _ _ => rfl) (fun _ _ => rfl) val_natCast
```

Comment:
Can you get any of these by generalizing the results in `Mathlib/Algebra/Order/Nonneg/Ring.lean`?